lib/src/adder.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// adder.dart
+// Implementation of Adder Module.
+//
+// 2023 June 1
+// Author: Yao Jing Quek <yao.jing.quek@intel.com>
+//
+
+import 'package:meta/meta.dart';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// An abstract class for all adder module.
+abstract class Adder extends Module {
+  /// The input to the adder pin [a].
+  @protected
+  final Logic a;
+
+  /// The input to the adder pin [b].
+  @protected
+  final Logic b;
+
+  /// The addition results [sum].
+  Logic get sum;
+
+  /// Takes in input [a] and input [b] and return the [sum] of the addition
+  /// result. The width of input [a] and [b] must be the same.
+  Adder(this.a, this.b, {super.name}) {
+    if (a.width != b.width) {
+      throw RohdHclException('inputs of a and b should have same width.');
+    }
+  }
+}
+
+/// A simple full-adder with inputs `a` and `b` to be added with a `carryIn`.
+class FullAdder extends Module {
+  /// The addition's result [sum].
+  Logic get sum => output('sum');
+
+  /// The carry bit's result [carryOut].
+  Logic get carryOut => output('carry_out');
+
+  /// Constructs a [FullAdder] with value [a], [b] and [carryIn] based on
+  /// full adder truth table.
+  FullAdder({
+    required Logic a,
+    required Logic b,
+    required Logic carryIn,
+    super.name = 'full_adder',
+  }) {
+    a = addInput('a', a, width: a.width);
```

Comment:
Move these `addInput`s for `a` and `b` into `Adder`, since you made the protected variables there.  That will also probably force them to be `late final` in `Adder`.

Reply:
Its seems like I need to removed the final.

```dart
/// An abstract class for all adder module.
abstract class Adder extends Module {
  /// The input to the adder pin [a].
  @protected
  late final Logic a;

  /// The input to the adder pin [b].
  @protected
  late final Logic b;

  /// The addition results [sum].
  Logic get sum;

  /// Takes in input [a] and input [b] and return the [sum] of the addition
  /// result. The width of input [a] and [b] must be the same.
  Adder(this.a, this.b, {super.name}) {
    if (a.width != b.width) {
      throw RohdHclException('inputs of a and b should have same width.');
    }
    a = addInput('a', a, width: a.width);
    b = addInput('b', b, width: b.width);
  }
}
```

Since `a` and `b` will be reinitialized.

Reply:
You should make it `late final`

Reply:
And don't make the inputs `this.a` and `this.b`, just `Logic a` and `Logic b`, then `this.a = addInput...`